### DISCUSSION TITLE
Arranging multiple blocks into a single block

### DISCUSSION THREAD

**Discussion by jmaas-onoffice** [WordPress/gutenberg#45413] [Q&A] [answered]
I'm trying to figure out a way to "combine" multiple blocks into one while having each "component" be independent, e.g. having its own toolbar.

For context: We are expecting that our customers would be overwhelmed by full Gutenberg with nesting blocks. For example, for adding an image next to a paragraph of text, we do not want them to have to use a columns block and then add the image left and the text right. Instead, we want to offer them a text-image block where they just need to choose the image and input the text.[^1]

We're getting quite close with Advanced Custom Fields blocks. However, in some preliminary user tests, we have seen the value of direct editing, especially directly editing text, which ACF blocks do not support.

The trouble I'm having right now is that Gutenberg is not expecting a block to be statically composed of other blocks. For example in the combined text-image block, I would like the image to be selectable like an independent block (similar to how it behaves in a core/columns).[^3]

The core/media-text this is "cheating" because the parent block controls are for the image and the paragraph is an InnerBlocks. So this does not work when you have more than two parts to be combined.[^2] For example imagine we would add a second image to core/media-text with the text in the middle. How could we offer controls for replacing each image separately?

[^1]: I'm aware there is a Media and Text block (core/media-text). I'm basically recreating it to test the concept and this is one of the simplest use cases for driving a prototype.
[^2]: Maybe I haven't delved deep enough into InnerBlocks. Is there a way to force specific inner blocks and control where they are placed?
[^3]: See https://github.com/WordPress/gutenberg/discussions/45413#discussioncomment-4018242 for a sketch of how this could look like.

**Comment by Mamaduka** [comment]
Hi, @jmaas-onoffice

I think creating a small wrapper block around core blocks is a good way to achieve desired results.

You'll need to use [InnerBlocks](https://developer.wordpress.org/block-editor/how-to-guides/block-tutorial/nested-blocks-inner-blocks/) with the combination of `allowedBlocks` and `template` properties. Then you can use CSS to handle the placement of the blocks.

**Reply by jmaas-onoffice** [reply]
But how would I arrange those into a single block? So imagine a block with an image on the left, a paragraph and a button below that:

<img width="932" src="https://user-images.githubusercontent.com/90763364/199007773-55aac19a-3cd4-463c-8f0c-7086001fa82b.png">

Instead of building this using a columns block, I want to offer one block with those three components and allow each component to be modified individually.

I hope this makes sense. 😅

**Reply by Mamaduka** [reply]
Have you considered creating a custom variation for the Columns block?

Learn more about block variations:

* https://developer.wordpress.org/block-editor/reference-guides/block-api/block-variations/
* https://fullsiteediting.com/lessons/block-variations/
* https://css-tricks.com/how-to-use-block-variations-in-wordpress/


## Example

```js
wp.blocks.registerBlockVariation (
  'core/columns', {
    name: 'media-example',
    title: 'Media Example',
    icon: 'smiley',
    scope: ['inserter'],
    innerBlocks: [
      ['core/column', {}, [
        ['core/image'],
      ]],
      ['core/column', {}, [
        ['core/paragraph', { placeholder: 'Content...' }],
      ]],
    ],
  }
);
```

**Reply by jmaas-onoffice** [reply]
Thanks for the pointers, I haven't looked into block variations before.

Unfortunately, I think they are not what I'm looking for, because they still expose the nested blocks to the user. That is, the user still needs to understand columns. We want to keep it really simple and want to hide a lot complexity and only expose specific controls. That's why we would like to build blocks ourselves, but oftentimes we want to group multiple components into a single block.

**Comment by bph** [comment, marked as answer]
another way to [curate the editor experience](https://developer.wordpress.org/block-editor/how-to-guides/curating-the-editor-experience/) for your content creators are outlined in the handbook, with [enhancements coming in WordPress 6.1](https://make.wordpress.org/core/2022/10/11/content-locking-features-and-updates/)

Block Pattern with locks might be your best and fastest solution, @jmaas-onoffice

**Reply by jmaas-onoffice** [reply]
Thanks for the pointers, I'll have a look!

**Reply by jmaas-onoffice** [reply]
I think I finally wrapped my head around it and this seems really close to what I'm looking for. 

With the "contentOnly" lock, I have my parent block using `InnerBlocks` with a template. In that template I can build up the layout using other Gutenberg blocks. The user will only be able to edit the content blocks such as images and text, not the layout blocks. So in that case it's basically what I was looking for.

I'm a bit skeptical where the limits of this are, so I'll have to play around with it a bit more and I'm not sure we will bet on this right now, but it's definitely an interesting direction that we will look out for.
